### PR TITLE
fix(types): provide better error messages for insights client

### PR DESCRIPTION
**Pull request by eunjae-lee**
## Summary

This PR updates the type `InsightsClient`. When the second parameter was wrong, TypeScript gave unhelpful error messages. With this change, now it can infer which method is being used, and thus provide better error messages about the second parameter.

fixes: #268

* sandbox: https://codesandbox.io/s/insights-typing-issue-with-clickedobjectidsaftersearch-forked-wzbon?file=/src/index.ts

---

### example 1

```js
insights("init", { appId, apiKeyy: "aaa" });
```

CleanShot 2021-05-26 at 11 03 51@2x

### example 2

```js
insights("clickedObjectIDsAfterSearch", {
  userToken: "user-1",
  eventName: "your_event_name",
  index: "your_index_name",
  // queryID: "queryID",
  objectIDs: ["objectID1", "objectID2"],
  positions: [17, 19]
});
```

CleanShot 2021-05-26 at 11 04 00@2x

### example 3

```js
insights('unknown_method');
```

CleanShot 2021-05-26 at 11 09 00@2x

Now it infers the method and provide error messages for that specific method.

### CODE REVIEW

File: lib/click.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { InsightsEvent } from "./_sendEvent";
 
 export interface InsightsSearchClickEvent {
+  eventName: string;
```
Comment (eunjae-lee) [inline lib/click.ts:4]: This was missing. It could be a separate PR on its own, but I just added it here.

File: lib/types.ts [hunk 2]
```diff
@@ -14,84 +14,63 @@ import {
 import { viewedObjectIDs, viewedFilters } from "./view";
 import { getVersion } from "./_getVersion";
 
-export type Init = (method: "init", ...args: Parameters<typeof init>) => void;
+export type InsightsMethodMap = {
```
Comment (eunjae-lee) [inline lib/types.ts:17]: Now we have a map of types of methods.

File: lib/types.ts [hunk 3]
```diff
@@ -14,84 +14,63 @@ import {
...
+type MethodType<MethodName extends keyof InsightsMethodMap> = (
+  method: MethodName,
+  ...args: InsightsMethodMap[MethodName]
 ) => void;
 
-export type AddAlgoliaAgent = (
-  method: "addAlgoliaAgent",
-  ...args: Parameters<typeof addAlgoliaAgent>
-) => void;
+export type Init = MethodType<"init">;
```
Comment (eunjae-lee) [inline lib/types.ts:41]: And each method is made from the map.

File: lib/types.ts [hunk 4]
```diff
@@ -14,84 +14,63 @@ import {
...
-  ConvertedObjectIDsAfterSearch &
-  ConvertedObjectIDs &
-  ConvertedFilters &
-  ViewedObjectIDs &
-  ViewedFilters;
+export type ViewedObjectIDs = MethodType<"viewedObjectIDs">;
+
+export type ViewedFilters = MethodType<"viewedFilters">;
+
+export type InsightsClient = <MethodName extends keyof InsightsMethodMap>(
```
Comment (eunjae-lee) [inline lib/types.ts:73]: `InsightsClient` takes the method name (which is one of the keys of the map) as the first parameter, and the second parameter is the exact parameter for the method.

File: lib/_getFunctionalInterface.ts [hunk 5]
```diff
@@ -7,6 +7,7 @@ export function getFunctionalInterface(
 ): InsightsClient {
   return (functionName, ...functionArguments) => {
     if (functionName && isFunction((instance as any)[functionName])) {
+      // @ts-ignore
```
Comment (shortcuts) [inline lib/_getFunctionalInterface.ts:10]: If this is meant to stay, we could use `@ts-expect-error` and explain why it's needed

File: lib/click.ts (hunk 1 above)
Comment (shortcuts) [reply lib/click.ts:4]: I think it's fine to have it here since it's still related to typescript

**Review by shortcuts — APPROVED** [review]
Looks good to me!

**Review by Haroenv — APPROVED** [review]
great change!
